python/ct/crypto/in_memory_merkle_tree.py

```diff
@@ -0,0 +1,125 @@
+"""In-memory Merkle Tree
+
+Operates (and owns) an in-memory array of leaves which can be updated.
+Not particularly efficient.
+"""
+
+import math
+
+from ct.crypto import merkle
+
+def _down_to_power_of_two(n):
+    """Returns the power-of-2 closest to n"""
+    if n < 2:
+        raise ValueError("N should be >= 2: %d" % n)
+    log_n = math.log(n, 2)
+    p = int(log_n)
+    # If n is exactly power of 2 then 2**p would be n, decrease p by 1.
+    if p == log_n:
+        p = p - 1
+    return 2**p
+
+class InMemoryMerkleTree(object):
+    """In-memory Merkle Tree representation. Not efficient or scalable."""
+
+    def __init__(self, leaves):
+        """Start with the array of |leaves| provided."""
+        self.__leaves = list(leaves)
+        self.__hasher = merkle.TreeHasher()
+
+    def _hashed_leaves(self):
+        """Returns an array of hashed leaves."""
+        return [self.__hasher.hash_leaf(t) for t in self.__leaves]
+
+    def add_leaf(self, leaf):
+        """Adds |leaf| to the tree, returning the index of the entry."""
+        self.__leaves.append(leaf)
+        return len(self.__leaves) - 1
+
+    def tree_size(self):
+        """Returns the size of the tree."""
+        return len(self.__leaves)
+
+    def get_root_hash(self, tree_size=None):
+        """Returns the root hash of the tree denoted by |tree_size|"""
```

Comment:
C6108: 44: One-line docstring summary should end with ".", "?", or "!" [g-short-docstring-punctuation]